chore: update platform controller with gcloud module

naitianliu-google opened the pull request:
#### Description
- add gcloud module in main.tf file to update platform controller for upgrading user cluster resource

#### Change summary
- add gcloud module in main.tf file
- declare cluster resource depends on the gcloud module
- update readme file for the instruction

akulkapoor90 commented on anthos-onprem-terraform/avmw_user_cluster_metallb/README.md:
```diff
@@ -58,20 +58,31 @@ have changed directory to where this samples is:
-admin cluster platform controller is listed in the
-[public documentation](https://cloud.google.com/anthos/clusters/docs/on-prem/latest/how-to/upgrade-on-prem-api#available_versions_for_upgrades).
+Before upgrading the user cluster, please make sure the admin cluster has been 
+enrolled in the Anthos On-Prem API. Steps for enrolling the admin cluster are
```
do we need to mention anything about the cloud controller here, since we are looking to launch that soon after which this step will become no-op.

naitianliu-google commented on anthos-onprem-terraform/avmw_user_cluster_metallb/main.tf:
```diff
@@ -47,12 +47,26 @@ resource "google_project_service" "default" {
+}
+
 # Create an anthos vmware user cluster and enroll it with the gkeonprem API
 resource "google_gkeonprem_vmware_cluster" "default" {
```
the 'default' naming is not a convention in terraform, we can use a different name here. We use 'default' here to keep is consistent with the resource name in ABM samples.

Shabirmean commented on anthos-onprem-terraform/avmw_user_cluster_metallb/main.tf:
```diff
@@ -47,12 +47,26 @@ resource "google_project_service" "default" {
   disable_on_destroy = false
 }
 
+# Update platform controller with gcloud command.
```
Can we add more descriptive comments here?
(suggestion)
```
# This module is used to upgrade the platform controller on your admin cluster. This
# is a necessary step for the user cluster version update. If the admin cluster is 
# already on the correct version, then this module does not change anything
```

Shabirmean replied:
I have used "upgrade" here to be consistent with the title. Please change it if `update` is more suitable

naitianliu-google replied:
Thanks for the suggestion. Added the suggested comments. I think it is more suitable to use `update` because it is also valid to update platform controller to an older version.

Shabirmean commented on anthos-onprem-terraform/avmw_user_cluster_metallb/main.tf:
```diff
@@ -47,12 +47,26 @@ resource "google_project_service" "default" {
 }
 
+# Update platform controller with gcloud command.
+module "gcloud-update-admin-cluster-platform-controller" {
```
```suggestion
module "gcloud_update_admin_cluster_platform_controller" {
```

Shabirmean replied:
The title state `upgrade`. Should we be consistent with that and use it instead of `update`? Or is it correct to call this module `glcoud_update_...`?

Shabirmean replied:
If it all should be upgrade, then please change all occurrences of update to upgrade!

Shabirmean commented on anthos-onprem-terraform/avmw_user_cluster_metallb/main.tf:
```diff
@@ -47,12 +47,26 @@ resource "google_project_service" "default" {
+  platform = "linux"
+
+  create_cmd_entrypoint = "gcloud"
+  create_cmd_body       = "beta container vmware admin-clusters update ${var.admin_cluster_name} --required-platform-version=${var.on_prem_version} --location ${var.region} --project ${var.project_id}"
```
```suggestion
  create_cmd_body       = <<EOT
    beta container vmware admin-clusters               \
    update ${var.admin_cluster_name}                   \
    --required-platform-version=${var.on_prem_version} \
    --project ${var.project_id}                        \
    --location ${var.region}                                           
  EOT
```

naitianliu-google replied:
Updated.

Shabirmean commented on anthos-onprem-terraform/avmw_user_cluster_metallb/README.md:
```diff
@@ -58,20 +58,31 @@ have changed directory to where this samples is:
-controller has been upgraded to the target version. The steps to upgrade the
-admin cluster platform controller is listed in the
-[public documentation](https://cloud.google.com/anthos/clusters/docs/on-prem/latest/how-to/upgrade-on-prem-api#available_versions_for_upgrades).
+Before upgrading the user cluster, please make sure the admin cluster has been 
```
_Please add the following description at the beginning:_

Use the same terraform script to upgrade the user cluster, by simply changing the version to the new version. Note that this script can be used for upgrades only if you had created the user cluster using this script. When you run the script with the updated version, the `terraform.tfstate` created during the first run of the script is compared to recognize the change.

Before upgrading the user cluster, please.....

naitianliu-google replied:
Added

Shabirmean commented on anthos-onprem-terraform/avmw_user_cluster_metallb/README.md:
````diff
@@ -58,20 +58,31 @@ have changed directory to where this samples is:
+   --location=REGION
 ```
 
+This sample shows how to update the platform controller with the following 
+gcloud module:
+
+- module `gcloud-update-admin-cluster-platform-controller`: You use this module
+  to upgrade the platform controller on the admin cluster. The platform
+  controller contains one or more bundles of components that the admin cluster
+  uses to manage user clusters. The bundles are version specific, that is, the
+  platform controller must contain a bundle version that matches the Anthos on
+  VMware version of the user cluster. Before upgrading your user cluster, run
+  this module to install a bundle version on the admin cluster that matches the
+  version you are upgrading the user cluster to.
+
````
I think we need to rephrase this a bit. The phrasing - "Before upgrading your user cluster, run this module" - makes it sound as if, this chunk of terraform can be independently run; which is not possible. Few things about this section:

- Use snake casing for the module name: `gcloud_update_admin_cluster_platform_controller`
- I would suggest the following re-write:

```suggestion
This `gcloud_update_admin_cluster_platform_controller` module uses the `gcloud`
command prepare the admin cluster to enable the user cluster upgrade.

- [**`gcloud-update-admin-cluster-platform-controller`**](./main.tf#L54-L62): 
   This module is used to ensure that the ** platform controller** of the admin cluster
   is on a compatible version. The platform controller contains one or more bundles of
   components that the admin cluster uses to manage user clusters. The bundles are
   version specific, that is, the platform controller must contain a bundle version that
   matches the _Anthos on VMware version of the user cluster_.  Thus, by having this
   module in the script we ensure that the platform controller in the admin cluster is 
   always on the correct user cluster version. 
```

naitianliu-google replied:
Updated. Thanks for the suggestion!

Shabirmean commented:
Please follow conventional style commits for the Conventional Commit CI test to pass. It should have the following style:

```
<type>: <simple message>

e.g:
fix: issue with the version
chore: cleanup the additonal lines
lint: fix the tab spaces
doc: add comments to describe resource
```

naitianliu-google commented:
@Shabirmean please help review the updated changes and let me know if anything needs to be corrected. Thanks!

Shabirmean commented on anthos-onprem-terraform/avmw_user_cluster_metallb/README.md:
```diff
@@ -58,20 +58,37 @@ have changed directory to where this samples is:
+This `gcloud_update_admin_cluster_platform_controller` module uses the `gcloud`
+command prepare the admin cluster to enable the user cluster upgrade.
+
+- [**`gcloud_update_admin_cluster_platform_controller`**](./main.tf#L54-L62): 
```
Now the line numbers are 53-67

```suggestion
- [**`gcloud_update_admin_cluster_platform_controller`**](./main.tf#L53-L67): 
```

naitianliu-google replied:
Updated. Thanks for the correction!

Shabirmean commented on anthos-onprem-terraform/avmw_user_cluster_metallb/main.tf:
```diff
@@ -47,12 +47,34 @@ resource "google_project_service" "default" {
+  version = "~> 2.0"
+
+  platform = "linux"
+
```
Remove unnecessary blank lines 56 and 58.

naitianliu-google replied:
Removed. Thanks for checking!

Shabirmean commented:
/gcbrun

naitianliu-google commented:
> LGTM! Please fix the two minor comments!
> 
> Thank you very much for addressing all the comments!

Updated accordingly. Thanks so much for taking your time to review this PR!

Shabirmean commented:
/gcbrun

Shabirmean commented:
@naitianliu-google -some minor static code analysis errors. Please fix 🙌🏼 

```
terraform fmt -recursive -check -diff ./
```

naitianliu-google commented:
> @naitianliu-google -some minor static code analysis errors. Please fix 🙌🏼
> 
> _To check_
> 
> ```
> terraform fmt -recursive -check -diff ./
> ```
> 
> _To fix_
> 
> ```
> terraform fmt -recursive ./
> ```

Corrected. Thanks for the guidance!

Shabirmean commented:
/gcbrun

naitianliu-google commented:
@Shabirmean please help take a look if this PR is ready to be merged or not. Thanks!
